### PR TITLE
feat: allow for dynamic hints

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ref: https://github.com/freeCodeCamp/freeCodeCamp/issues/43668

### CODE REVIEW

**Review by ahmaxed — APPROVED** [review]

**Review by ShaunSHamilton — COMMENTED** [review]
I like this approach, and think this will be fun to make tests with.

However, I wonder if it would not be better to keep the current format, and just insert the assertion under the text:
image

File: client/src/templates/Challenges/redux/execute-challenge-saga.js [hunk 1]
```diff
@@ -167,7 +167,11 @@ function* executeTests(testRunner, tests, testTimeout = 5000) {
         throw err;
       }
     } catch (err) {
-      newTest.message = text;
+      const { actual, expected } = err;
+
+      newTest.message = text
+        .replace('--fcc-expected--', expected)
+        .replace('--fcc-actual--', actual);
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/redux/execute-challenge-saga.js:174]: The above just looks like this:
```suggestion
      newTest.message = `${text}\n\n${message}`;
```

Comment (ShaunSHamilton) [reply client/src/templates/Challenges/redux/execute-challenge-saga.js:174]: Either that, or simply the following which can easily be translated:
```suggestion
      const translatableAssertion = `Expected ${expected}, but found ${actual}`;
      newTest.message = `${text}\n\n{translatableAssertion}`;
```

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/redux/execute-challenge-saga.js:174]: Thing is, what happen if you don't want to say exactly "Expected ..., but found ..."?  

Also, that locks into either outputting this for every test or adding more code (frontmatter, maybe?) to configure which tests show it.

For what it's worth
```
Your `h1` element should have the text `Hello World`.
Expected --fcc-expected--, but found --fcc-actual--. 
```

works with the current parser.

**Comment by ojeytonwilliams** [comment]
Ooops:

image

Let me fix that.

File: client/src/templates/Challenges/redux/execute-challenge-saga.js (hunk 1 above)
Comment (ShaunSHamilton) [reply client/src/templates/Challenges/redux/execute-challenge-saga.js:174]: > Thing is, what happen if you don't want to say exactly "Expected ..., but found ..."?

This is why I would prefer sticking with relying on the actual assertion message from specific Chai methods.

> more code (frontmatter, maybe?)

Maybe? Either `--fcc-actual--...` or some flag at the end of a hint `--fcc-show-assertion--`

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/redux/execute-challenge-saga.js:174]: `--fcc-show-assertion--`, yeah, that's where my mind went.  I think that's a reasonable approach.

**Review by ShaunSHamilton — APPROVED** [review]
This works, and LGTM 👍 

We will iron out implementations of this, as we go on. Thanks, Oliver.

**Pull request merged**